### PR TITLE
Allow hostports with localhost hostIPs

### PR DESCRIPTION
**Pull request by aspsk**
We need to enable localhost addresses for hostIP, because this is one of the reasons why the following conformance test fails:

  sonobuoy run --e2e-focus 'validates that there is no conflict between pods with same hostPort but different hostIP and protocol'

And we need all conformance tests to pass.  (This is not the only reason, the other one is that we don't distinguish between TCP and UDP services, yet.)

The patch is pretty straightforward (we previously just didn't let to specify localhost addresses for hostports in the pod initialization routine).  One note, however, is that we need to explicitly check that pods are not sending data to such "localhost" hostports - only the root namespace should be allowed to connect. We need to do this in both bpf_lxc.c and bpf_sock.c for per-connection and per-packet service translations, correspondingly.

```release-note
Allow users to specify hostports with localhost hostIP
```

### CODE REVIEW

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -1258,6 +1262,10 @@ static __always_inline int __tail_handle_ipv4(struct __ctx_buff *ctx)
 				goto skip_service_lookup;
 			}
 #endif /* ENABLE_L7_LB */
+
+			if (is_v4_loopback(ip4->daddr))
+				goto skip_service_lookup;
+
```
Comment (borkmann) [inline bpf/bpf_lxc.c]: I think this is not enough given also traffic on the wire could get to the node where we have the 127.0.0.1 as IP in svc table, so also this needs to be rejected from devs with bpf_host/xdp on it.

I wonder.. are hostport ranges forbidden to clash with nodeport ranges? If so, maybe we could solve this issue by installing a wildcard entry, which covers loopback and all node local addresses - similar as we do in kpr for nodeport.

The logic in sock LB, should probably go into sock{4,6}_skip_xlate() which already has checks around lb{4,6}_svc_is_hostport().

Comment (aspsk) [reply bpf/bpf_lxc.c]: Thanks @borkmann !

Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

For the port range: I believe that the hostPort range is not defined, so it can be any, but the NodePort range is reserved, so we won't let users to create HostPorts within the NodePort range. As a matter of fact, I've just checked that you added this check in 2020 in a commit named "prep datapath for loopback/local address exposure on hostport", so I will look into installing the wildcard entry :-)

Comment (borkmann) [reply bpf/bpf_lxc.c]: > Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

Yes, this exactly, e.g. you could craft af_packet skb and push it to the wire when in same L2.

**Comment by aspsk** [comment]
/test-1.16-4.9

**Comment by aspsk** [comment]
/test-1.25-net-next

**Review by joamaki — APPROVED** [review]
Looks good. General observations: the whole wildcard thing is a bit hard to follow. Maybe the lookup functions could be better documented to show what are the concrete lookups they perform? We should perhaps also consider using bit flags with named constants when we've got more than one boolean flag, e.g. `sock4_wildcard_lookup(key, WCL_INV_MATCH | WCP_IN_HOSTNS)`.

I would recommend adding a test to `bpf/tests` for this as it should not be too difficult now and a good idea to extend our still somewhat poor coverage whenever we touch the bpf code. `host_only_socket_lb_test.c` is probably closest as a starting point.

**Comment by aspsk** [comment]
> Looks good. General observations: the whole wildcard thing is a bit hard to follow. Maybe the lookup functions could be better documented to show what are the concrete lookups they perform?

Thanks, I will add some words

> We should perhaps also consider using bit flags with named constants when we've got more than one boolean flag, e.g. `sock4_wildcard_lookup(key, WCL_INV_MATCH | WCP_IN_HOSTNS)`.

I was also thinking of adding named flags (not only here, we have many `foo(..., true, false, ...)` calls). Though I would leave individual arguments vs. or'ed flags.

> I would recommend adding a test to `bpf/tests` for this as it should not be too difficult now and a good idea to extend our still somewhat poor coverage whenever we touch the bpf code. `host_only_socket_lb_test.c` is probably closest as a starting point.

Thanks, I will add a test.

File: pkg/k8s/watchers/pod.go [hunk 2]
```diff
@@ -479,6 +483,11 @@ func (k *K8sWatcher) deleteK8sPodV1(pod *slim_corev1.Pod) error {
 	return err
 }
 
+func netnsCookieSupported() bool {
+	return probes.HaveProgramHelper(ebpf.CGroupSock, asm.FnGetNetnsCookie) == nil &&
```
Comment (joamaki) [inline pkg/k8s/watchers/pod.go]: This will crash tests as they cannot perform this syscall (they're not running as root, or the tests are run under macOS). This should be resolved at daemon initialization and provided via e.g. `option.Config` or somehow passed to k8s watcher. Or can we actually just drop the whole `feIP` check completely? If it would only affect old kernels I don't think there's much value to it if we need to go out of our way to supply the probe result to this method.

The real problem here is that genServiceMappings really doesn't belong in pkg/k8s/watchers. It's something that should be somewhere under pkg/datapath. That's unfortunately bigger refactoring.

**Comment by aspsk** [comment]
/test-1.25-net-next

**Comment by aspsk** [comment]
/test-1.23-5.4

**Comment by sayboras** [comment]
Removed myself from the review list, as we already have expertise from other reviewers, and I am unfamiliar with this code path.

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (aspsk) [reply pkg/k8s/watchers/pod.go]: Tests shoudn't break, as we don't test this functionality from user space

**Review by borkmann — APPROVED** [review]

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (brb) [inline pkg/k8s/watchers/pod.go]: Shouldn't this be wrapped into `sync.Once()`?

Comment (brb) [reply pkg/k8s/watchers/pod.go]: > The real problem here is that genServiceMappings really doesn't belong in pkg/k8s/watchers.

ACK. Just that, unfortunately, at the moment I don't see any way around than extending the refactoring backlog.

File: bpf/bpf_sock.c [hunk 3]
```diff
@@ -226,20 +226,29 @@ static __always_inline struct lb4_service *
...
+
+	/*
+	 * We perform a wildcard hostport lookup. If the SVC_FLAG_LOOPBACK
+	 * flag is set, then this means that the wildcard entry was set up
+	 * using a loopback IP address, in which case we only want to allow
+	 * connections to loopback addresses
+	 */
+	key->address = orig_addr;
+	svc = sock4_wildcard_lookup(key, false, true, in_hostns);
+	if (svc && lb4_svc_is_hostport(svc) && (!lb4_svc_is_loopback(svc) || loopback))
```
Comment (brb) [inline bpf/bpf_sock.c:247]: Why don't we use the following condition instead?

`(svc && lb4_svc_is_hostport(svc) && lb4_svc_is_loopback(svc) && loopback)`

**Review by brb — COMMENTED** [review]
LGTM, just a few questions / comments.

File: bpf/bpf_sock.c (hunk 3 above)
Comment (aspsk) [reply bpf/bpf_sock.c:247]: @brb This is a different condition ^ it will reject all lookups for non-`SVC_FLAG_LOOPBACK` services. Does the comment in the code above makes sense to you?

**Comment by T-Kukawka** [comment]
Hi, when  could we expect this PR to be merged and released? this blocks our releases due to the CNCF tests which are not passing  since september :(

File: bpf/bpf_sock.c (hunk 3 above)
Comment (brb) [reply bpf/bpf_sock.c:247]: Yep, it does. Thanks.

**Review by brb — APPROVED** [review]
:rocket:

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (joamaki) [reply pkg/k8s/watchers/pod.go]: The probing result is memoized (in cilium/ebpf's HaveProgramHelper), so sync.Once isn't really needed. This is probably a fairly recent change.

**Comment by aspsk** [comment]
> Hi, when could we expect this PR to be merged and released? this blocks our releases due to the CNCF tests which are not passing since september :(

Hi @T-Kukawka This PR, hopefully, will be merged tomorrow.

Unfortunately, this won't fix the CNCF test, as there's also the following issue: https://github.com/cilium/cilium/issues/9207 I will look into it next, however, I am not sure that it will be resolved right away (there were several attempts to resolve it already, if you follow the logs in the ticket)

**Comment by T-Kukawka** [comment]
@aspsk thanks for the update!

However this is great blocker for us and issue you referenced is 2 years old. Does that mean that cilium is not CNCF compatible at all when running without kube-proxy?  Shall we drop it? We have been postponing that release for a while already and this does not look promising at all. We need the CNCF certifications.

**Comment by aspsk** [comment]
> However this is great blocker for us and issue you referenced is 2 years old. Does that mean that cilium is not CNCF compatible at all when running without kube-proxy? Shall we drop it? We have been postponing that release for a while already and this does not look promising at all. We need the CNCF certifications.

Hi @T-Kukawka 

Yes, with kube-proxy replacement the CNCF test fails due to either the "TCP vs. UDP" bug or the "hostport on localhost" bug (fixed by this commit). This wasn't a big priority for us, because on practice backends tend to run both TCP and UDP services, coredns as an example. (It fails only if one backend runs TCP and the other UDP on the same port.)

I will look into the "TCP vs. UDP" thing soon and will update you as soon as we have a plan and/or patches to test.

**Comment by T-Kukawka** [comment]
thanks  @aspsk! highly appreciated :)

**Comment by aspsk** [comment]
/test-1.16-4.9

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (aspsk) [reply pkg/k8s/watchers/pod.go]: Let me keep it @joamaki, I've never used `sync.Once` before :pray: :rofl:

Comment (aspsk) [reply pkg/k8s/watchers/pod.go]: @joamaki Actually, I think this is really ok to leave the `sync.Once` here. Even though the `HaveProgramHelper` (more precisely the `ebpf.features.haveProgramHelper`) caches the result, this should be some cycles faster with the `sync.Once`. So I will mark as resolved, feel free to, revert

**Comment by whites11** [comment]
@aspsk what version of cilium should we expect this PR to land in? just 1.13 or is it any backport planned? :bowing_man:

**Comment by aspsk** [comment]
> @aspsk what version of cilium should we expect this PR to land in? just 1.13 or is it any backport planned? bowing_man

Hi @whites11 I would expect the 1.13 only, if no one of our paid customers require it to be backported
